libcxx/include/__config

```diff
@@ -280,6 +280,14 @@
 // - `_LIBCPP_ASSERT_NON_OVERLAPPING_RANGES` -- for functions that take several ranges as arguments, checks that the
 //   given ranges do not overlap.
 //
+// - `_LIBCPP_ASSERT_VALID_DEALLOCATION` -- checks that an attempt to deallocate memory is valid (e.g. the given object
+//   was allocated by the given allocator). Violating this category typically results in a memory leak.
+//
+// - `_LIBCPP_ASSERT_VALID_EXTERNAL_API_CALL` -- checks that a call to an external API (e.g. a syscall) doesn't fail in
+//   an unexpected manner. This includes triggering documented cases of undefined behavior in an external library (like
+//   attempting to unlock an unlocked mutex in pthreads). We generally don't expect these failures to compromize memory
+//   safety or otherwise create an immediate security issue.
+//
```

Comment:
Not specific to this commit, but I feel this list is getting a bit long in here. How about putting it in an `.rst` file. I also would like a better overview what the modi do. For example, what does `_LIBCPP_HARDENING_MODE_FAST`  do.
Maybe a summary lke
```
* _LIBCPP_HARDENING_MODE_FAST
  * foo
  * bar
* _LIBCPP_HARDENING_MODE_EXTENSIVE
   * all of _LIBCPP_HARDENING_MODE_FAST
   * do
   * re
   * mi
```

Reply:
I completely agree, and I plan to improve the documentation after this round of categorization patches lands.

Reply:
That probably happens after branching LLVM 18. After branching we can always backport documentation changes. So it would be great if we can get better documentation in LLVM 18.